velox/common/memory/ArbitrationOperation.h

```diff
@@ -0,0 +1,171 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/GTestMacros.h"
+#include "velox/common/base/StatsReporter.h"
+#include "velox/common/future/VeloxPromise.h"
+#include "velox/common/memory/ArbitrationParticipant.h"
+#include "velox/common/memory/Memory.h"
+
+namespace facebook::velox::memory {
+
+/// Manages the execution of a memory arbitration request within the arbitrator.
+class ArbitrationOperation {
+ public:
+  ArbitrationOperation(
+      ScopedArbitrationParticipant&& pool,
+      uint64_t requestBytes,
+      uint64_t timeoutMs);
+
+  ~ArbitrationOperation();
+
+  enum class State {
+    kInit,
+    kWaiting,
+    kRunning,
+    kFinished,
+  };
+  State state() const {
+    return state_;
+  }
+  static std::string stateName(State state);
+
+  /// Returns the corresponding arbitration participant.
+  const ScopedArbitrationParticipant& participant() {
+    return participant_;
+  }
+
+  /// Invoked to start arbitration execution on the arbitration participant. The
+  /// latter ensures the serialized execution of arbitration operations from the
+  /// same query with one at a time. So this method blocks until all the prior
+  /// arbitration operations finish.
+  void start();
+
+  /// Invoked to finish arbitration execution on the arbitration participant. It
+  /// also resumes the next waiting arbitration operation to execute if there is
+  /// one.
+  void finish();
+
+  /// Returns true if the corresponding arbitration participant has been
+  /// aborted.
+  bool aborted() const;
+
+  /// Invoked to set the grow targets for this arbitration operation based on
+  /// the request size.
+  ///
+  /// NOTE: this should be called once after the arbitration operation is
+  /// started.
+  void setGrowTargets();
+
+  uint64_t requestBytes() const {
+    return requestBytes_;
+  }
+
+  /// Returns the max grow bytes for this arbitration operation which could be
+  /// larger than the request bytes for exxponential growth.
+  uint64_t maxGrowBytes() const {
+    return maxGrowBytes_;
+  }
+
+  /// Returns the min grow bytes for this arbitration operation to ensure the
+  /// arbitration participant has the minimum amount of memory capacity. The
+  /// arbitrator might allocate memory from the reserved memory capacity pool
+  /// for the min grow bytes.
+  uint64_t minGrowBytes() const {
+    return minGrowBytes_;
+  }
+
+  /// Returns the allocated bytes by this arbitration operation.
+  uint64_t& allocatedBytes() {
+    return allocatedBytes_;
```

Comment:
just make it public or protected if needed to be accessed like this?

Reply:
Yes, in followup.